docetl/operations/code_operations.py

```diff
@@ -0,0 +1,107 @@
+from typing import Any, Dict, List, Optional, Tuple
+from concurrent.futures import ThreadPoolExecutor
+from docetl.operations.base import BaseOperation
+from docetl.operations.utils import RichLoopBar
+from pydantic import Field
+
+class CodeMapOperation(BaseOperation):
+    class schema(BaseOperation.schema):
+        type: str = "code_map"
+        code: str
+        drop_keys: Optional[List[str]] = None
+
+    def syntax_check(self) -> None:
+        config = self.schema(**self.config)
+        try:
+            namespace = {}
+            exec(config.code, namespace)
+            if "transform" not in namespace:
+                raise ValueError("Code must define a 'transform' function")
+            if not callable(namespace["transform"]):
+                raise ValueError("'transform' must be a callable function")
+        except Exception as e:
+            raise ValueError(f"Invalid code configuration: {str(e)}")
+
+    def execute(self, input_data: List[Dict]) -> Tuple[List[Dict], float]:
+        namespace = {}
+        exec(self.config["code"], namespace)
+        transform_fn = namespace["transform"]
```

Comment:
Let's merge the operation result with the original document, so we don't lose keys as we do operations. Example [here](https://github.com/ucbepic/docetl/blob/75e62c9677fd985a9d1e7c9567b2b5ced84e59c3/docetl/operations/map.py#L231) in the original map operation

Reply:
done

Reply:
```
def execute(self, input_data: List[Dict]) -> Tuple[List[Dict], float]:
        namespace = {}
        exec(self.config["code"], namespace)
        transform_fn = namespace["transform"]

        results = []
        with ThreadPoolExecutor() as executor:
            futures = [executor.submit(transform_fn, doc) for doc in input_data]
            pbar = RichLoopBar(
                range(len(futures)),
                desc=f"Processing {self.config['name']} (code_map)",
                console=self.console,
            )
            for i, doc in enumerate(input_data):
                result = futures[i].result()
                if self.config.get("drop_keys"):
                    result = {
                        k: v for k, v in result.items() 
                        if k not in self.config["drop_keys"]
                    }
                merged_result = {**doc, **result}
                results.append(merged_result)
        return results, 0.0
```